### ISSUE TITLE
function in opted out library ran from opted-in entry point is run as weak (expect failure!)

### ISSUE THREAD

**Issue by kevmoo** [dart-lang/sdk#41903]
See https://github.com/dart-lang/null_safety_sample_package

```console
> dart bin/main.dart
Case 1 /Users/kevmoo/homebrew/Cellar/dart/2.9.0-9.0.dev/libexec/bin/dart bin/strong_entry.dart
opted-in entry point is weak? true
  func in opted-out lib is weak? true
  func in opted-in lib is weak? true
# As expected!

Case 2 /Users/kevmoo/homebrew/Cellar/dart/2.9.0-9.0.dev/libexec/bin/dart bin/weak_entry.dart
opted-out entry point is weak? true
  func in opted-out lib is weak? true
  func in opted-in lib is weak? true
# As expected!

Case 3 /Users/kevmoo/homebrew/Cellar/dart/2.9.0-9.0.dev/libexec/bin/dart --enable_experiment=non-nullable bin/strong_entry.dart
opted-in entry point is weak? false
# expect this to fail! An opted-in entry point importing and running a function in an opted-out library should fail, right?
  func in opted-out lib is weak? true
  func in opted-in lib is weak? false
# not expected!

Case 4 /Users/kevmoo/homebrew/Cellar/dart/2.9.0-9.0.dev/libexec/bin/dart --enable_experiment=non-nullable bin/weak_entry.dart
opted-out entry point is weak? true
  func in opted-out lib is weak? true
  func in opted-in lib is weak? true
# As expected!
```

**Comment by leafpetersen** [comment]
@kevmoo Some confusion happening here.

1) we don't plan to support mixed mode + strong mode execution.   @johnniwinther is about to land this restriction.  That makes both of your strong_entry examples static errors unless you run --no-sound-null-safety to disable strong checking.
2) I don't understand why the first two examples work without enabling the experiment.

There is a known bug with const evaluation, I don't think it applies here but it might.  Removing the `const` will allow you to see if that's what you're hitting.  I think what you want to be testing is:

1) A strong entry point that imports a strong library gets run in strong mode
2) A strong entry point that imports a weak library gets run in strong mode and emits a compile error
3) A strong entry point that imports a weak library gets run with --no-sound-null-safety correctly runs in weak mode
4) a weak entry point that imports a strong library gets run in weak mode
5) A weak entry point that imports a weak library gets run in weak mode
6) A weak entry point that gets run with --sound-null-safety gets a compile error

**Comment by a-siva** [comment]
I tried your example above, opting out bin/weak_lib.dart causes the entire program to run in weak mode. @johnniwinther per comments from Leaf above it looks like that case should be returning strong mode which will later lead to a compile error when we invoke the compiler.

**Comment by kevmoo** [comment]
I talked to Leaf about this. Running with the SDK @ head things are working
as expected – we think.

I'm working on an updated thing to check things here...

**Comment by kevmoo** [comment]
@johnniwinther fixed this in 00353040b9034217962c80947d7b064ef28767d7 – we're all good now!
